File: Task3/15/zhangyongshun.m
xdat = load('q1x.dat');
ydat = load('q1y.dat');
col = length(ydat);
xdat = [ones(col,1) xdat];



for i = 1:col               %show the points(x1,x2) with diffrent symbols('o' represent 'y = 1'and '+' represent 'y = 0') 
    if ydat(i)
        plot(xdat(i,2),xdat(i,3),'ro');
        hold on;
    else
        plot(xdat(i,2),xdat(i,3),'b+');
        hold on;
    end;
end;

theta = [1;1;1];              %initialise matrix theta and learning rate alf
alf = 1;

while 1                     %gradient descent
    h = 1 ./ (1 + exp(-xdat * theta));
    tmp_theta = theta - (1 / col) * alf * xdat' * (h - ydat);
    if abs(sum(tmp_theta - theta)) <= 0.1
        theta = tmp_theta;
        break;
    end;
    theta = tmp_theta;
end;

xdat = [min(xdat(:,2)) - 3 max(xdat(:,2) + 3)];
ydat = [ -1. / theta(3) * (theta(2) * xdat + theta(1))];
plot(xdat,ydat);
hold off;

title('ZhangYongShun Task3');

print -dpng  'zhangyongshun.png'

File: Task3/15/zhangyongshun.py
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 03 20:21:24 2016

@author: Xiao顺
"""

import numpy as np
import matplotlib.pyplot as plt  

x = np.mat(np.loadtxt("D:\python_files\q1x.dat",dtype = float))
y =  np.mat(np.loadtxt('D:\python_files\q1y.dat',dtype = float).ravel()).T
col = np.size(x,0);
tmp = np.mat(np.ones(col)).T;
x = np.bmat("tmp x")

for i in range(0,col):
    tmp = 'ro' if int(y[i,0])  else 'r+'
    plt.plot(x[i,1],x[i,2],tmp)


theta = np.mat(np.ones(3).reshape(3,1))
alf = 1

while True:
    h = 1. / (1 + np.exp(-x * theta))
    tmp_theta = theta - (1. / col) * alf * x.T * (h - y)
    if abs(sum(tmp_theta - theta)) <= 0.1:
        theta = tmp_theta
        break
    theta = tmp_theta

X= [int(min(x[:,1]) - 3) ,int(max(x[:,1]) + 3)]
Y = [-1. / theta[2,0] * (theta[1,0] * X[0] + theta[0,0]),-1. / theta[2] * (theta[1] * X[1] + theta[0])]
plt.plot(X,Y)

plt.title('ZhaangYongShun Task3')



  